File: scalether/util/src/main/java/scalether/util/Bytes.java
package scalether.util;

import java.util.Arrays;

public class Bytes {
    public static final byte ZERO = 0x0;
    public static final byte ONE = 0x1;

    private Bytes() {}

    public static void fill(byte[] bytes, byte fill, int from, int to) {
        for (int i = from; i < to; i++) {
            bytes[i] = fill;
        }
    }

    public static byte[] filled(int size, byte fill) {
        final byte[] result = new byte[size];
        if (fill != ZERO) {
            fill(result, fill, 0, size);
        }
        return result;
    }

    public static byte[] trimLeadingBytes(byte[] bytes, byte b) {
        int offset = 0;
        for (; offset < bytes.length - 1; offset++) {
            if (bytes[offset] != b) {
                break;
            }
        }
        return Arrays.copyOfRange(bytes, offset, bytes.length);
    }

    public static byte[] trimLeadingZeroes(byte[] bytes) {
        return trimLeadingBytes(bytes, (byte) 0);
    }

}

File: domain/src/main/scala/io/daonomic/rpc/domain/Bytes.scala
package io.daonomic.rpc.domain

import java.math.BigInteger
import java.util

import scalether.util.{Hex, Padding}

trait Bytes {
  def bytes: Array[Byte]

  def slice(from: Int, until: Int): Binary =
    Binary(bytes.slice(from, until))

  def add(other: Bytes): Binary = {
    Binary(bytes ++ other.bytes)
  }

  def add(other: Array[Byte]): Binary = {
    Binary(bytes ++ other)
  }

  def ++(other: Bytes): Binary = add(other)

  def ++(other: Array[Byte]): Binary = add(other)

  def length: Int = bytes.length

  def padded: Binary =
    Binary.apply(Padding.padLeft(bytes, scalether.util.Bytes.ZERO))

  def toBigInteger: BigInteger =
    new BigInteger(Hex.to(bytes), 16)

  override def toString: String =
    Hex.prefixed(bytes)

  def prefixed: String =
    toString()

  def hex: String =
    Hex.to(bytes)

  def toBinary: Binary

  override def equals(obj: scala.Any): Boolean = {
    if (!obj.isInstanceOf[Bytes]) {
      false
    } else {
      util.Arrays.equals(obj.asInstanceOf[Bytes].bytes, bytes)
    }
  }

  override def hashCode(): Int = bytes.toSeq.hashCode()
}
